src/netlink.c

```diff
@@ -0,0 +1,365 @@
+\*=========================================================================*/
+static int global_create(lua_State *L);
+static int meth_bind(lua_State *L);
+static int meth_sendto(lua_State *L);
+static int meth_send(lua_State *L);
+static int meth_receivefrom(lua_State *L);
+static int meth_receive(lua_State *L);
+static int meth_close(lua_State *L);
+static int meth_settimeout(lua_State *L);
+static int meth_gettimeout(lua_State *L);
+static int meth_getfd(lua_State *L);
+static int meth_setfd(lua_State *L);
+static int meth_getpeername(lua_State *L);
+static int meth_setpeername(lua_State *L);
+static int meth_getsockpid(lua_State *L);
+
+static const char *netlink_trybind(p_netlink nl, int grp);
+
+/* netlink object methods */
+static luaL_Reg netlink_methods[] = {
+    {"__gc",        meth_close},
+    {"__tostring",  auxiliar_tostring},
+    {"bind",        meth_bind},
+    {"close",       meth_close},
+    {"getfd",       meth_getfd},
+    {"send",        meth_send},
+    {"sendto",      meth_sendto},
+    {"receivefrom", meth_receivefrom},
+    {"receive",     meth_receive},
+    {"setfd",       meth_setfd},
+    {"settimeout",  meth_settimeout},
+    {"gettimeout",  meth_gettimeout},
+    {"setpeername", meth_setpeername},
+    {"getpeername", meth_getpeername},
+    {"getsockpid",  meth_getsockpid},
+    {NULL,          NULL}
+};
+
+/* functions in library namespace */
+static luaL_Reg func[] = {
+    {"netlink", global_create},
+    {NULL, NULL}
+};
+
+/*-------------------------------------------------------------------------*\
+* Initializes module
+\*-------------------------------------------------------------------------*/
+int netlink_open(lua_State *L) {
+    /* create classes */
+    auxiliar_newclass(L, "netlink{connected}", netlink_methods);
+    auxiliar_newclass(L, "netlink{unconnected}", netlink_methods);
+    /* create class groups */
+    auxiliar_add2group(L, "netlink{connected}", "netlink{any}");
+    auxiliar_add2group(L, "netlink{unconnected}", "netlink{any}");
+    luaL_setfuncs(L, func, 0);
+    return 0;
+}
+
+/*=========================================================================*\
+* Lua methods
+\*=========================================================================*/
+/*-------------------------------------------------------------------------*\
+* Send data through connected netlink socket
+\*-------------------------------------------------------------------------*/
+static int meth_send(lua_State *L) {
+    p_netlink nl = (p_netlink)auxiliar_checkclass(L, "netlink{connected}", 1);
+    size_t payload_size;
+    const char *payload = luaL_checklstring(L, 2, &payload_size);
+    int flags = luaL_optinteger(L, 3, 0);
+    struct nlmsghdr nlh[NLMSG_SPACE(payload_size)];
+    p_timeout tm = &nl->tm; 
+    size_t sent = 0;
+    int err;
+
+    memset(nlh, 0, NLMSG_SPACE(payload_size));
+    nlh->nlmsg_len = NLMSG_SPACE(payload_size);
+    nlh->nlmsg_pid = nl->srcpid;
+    nlh->nlmsg_flags = flags;
```

Comment:
How about other header fields like `nlmsg_seq` and `nlmsg_type`? The application cannot use it? We're not using it in our protocol over Netlink to talk to out code in the kernel?

Reply:
He would have to find out the numeric values, however it's an optional field. If he wanted to specify a flag he could, but it's not necessary. 
Why would that flag render the message illegal? 
nlmsg_seq and nlmsg_type are opaque to the netlink core. We could use them, but they would only serve a purpose if we assigned them a purpose on our implementation.

Reply:
Why are you telling me exactly the problems I've asked you to address?
Did you bother to read the docs about `NLM_F_MULTI`?
Again, why are you telling me what `nlmsg_seq` and `nlmsg_type` are instead of addressing my point about the application and out code in the kernel not being able to use it?